Lib/test/test_os.py

```diff
@@ -2873,6 +2873,48 @@ def test_getfinalpathname_handles(self):
 
         self.assertEqual(0, handle_delta)
 
+    @support.requires_subprocess()
+    def test_stat_unlink_race(self):
+        # bpo-46785: the implementation of os.stat() falls back to reading
+        # the parent directory if CreateFileW() fails with a permission
+        # error. If reading the parent directory fails because the file or
+        # directory are subsequently unlinked, or because the volume or
+        # share are no longer available, then the original permission error
+        # should not be restored.
+        filename =  os_helper.TESTFN
+        self.addCleanup(os_helper.unlink, filename)
+        command = '''if 1:
+            import os
+            import sys
+            filename = sys.argv[1]
+            while True:
+                try:
+                    with open(filename, "w") as f:
+                        pass
+                except OSError:
+                    pass
+                try:
+                    os.remove(filename)
+                except OSError:
+                    pass
+        '''
+        deadline = time.time() + 5
+        p = subprocess.Popen([sys.executable, '-c', command, filename])
```

Comment:
Use this in a `with` block. (Particularly in tests, cleanup of resources is important.)

Reply:
There's nothing to be cleaned up by the context manager here. In Windows, the garbage collector handles closing the `Handle` for the process lazily. That's fine since the process is already rundown in the kernel. What's left of the kernel process object is just for bookkeeping (e.g. exit status, creation time, exit time, etc). There's no requirement to wait on a child process in Windows.

Also, while it doesn't apply here, note that there's fragility in the design of `subprocess.Popen` that can hang in `__exit__()`.  See [bpo-43346](https://bugs.python.org/issue43346). WinAPI `CloseHandle()` for a file is synchronous when it's the last handle for the file in the current process, since the I/O manager needs to clean up some process context, such as byte range file locking. If a thread is blocking on a synchronous `read()` or `write()` of the file, calling `close()` in `Popen.__exit__()` will thus block until that I/O operation completes. Generally that's okay, but pipe handles are often leaked to descendant processes, particularly with `shell=True`, which keeps the pipe open even after the child process is terminated.

Reply:
Done.

Reply:
@eryksun I missed your comment earlier. I'm neutral on this, and since both ways are used in many places in the tests, I'd say that either option should be OK.

Reply:
> There's nothing to be cleaned up by the context manager here.

Today, but it may well be added one day. Possibly even resource tracking could be added. Better to use the API as intended even if it isn't strictly necessary, to protect against future changes.